### PR TITLE
Write a bunch of libc tests, and reimplemented a bunch while I was at it.

### PR DESCRIPTION
**Pull request by paladine**

### CODE REVIEW

File: MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs [hunk 1]
```diff
@@ -4327,29 +4312,10 @@ private void stpbtvl()
...
-                if (string1[i] == string2[i]) continue;
-
-                //1 < 2 == -1 (0xFFFF)
-                //1 > 2 == 1 (0x0001)
-                Registers.AX = (ushort)(string1[i] < string2[i] ? 0xFFFF : 1);
-                return;
-            }
-
-            Registers.AX = 0;
+            Registers.AX = (ushort)(short)String.Compare(string1, string2);
```
Comment (enusbaum) [inline MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs:4318]: `string` vs `String` alias
Unnecessary cast to `(short`) before `(ushort)`. `int` -1 will cast to `ushort` 0xFFFF

image

File: MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs [hunk 2]
```diff
@@ -4893,47 +4859,52 @@ private void gabbtv()
...
 
-            if (!Module.Memory.TryGetVariablePointer("STROK-RESULT", out var resultPointer))
-                resultPointer = Module.Memory.AllocateVariable("STROK-RESULT", 0x400);
-
-            if (!Module.Memory.TryGetVariablePointer("STROK-ORIGINAL", out var originalPointer))
-                originalPointer = Module.Memory.AllocateVariable("STROK-ORIGINAL", 0x400);
-
-            if (!Module.Memory.TryGetVariablePointer("STROK-ORDINAL", out var ordinalPointer))
-                ordinalPointer = Module.Memory.AllocateVariable("STROK-ORDINAL", 0x2);
+            var workPointerPointer = Module.Memory.GetOrAllocateVariablePointer("STRTOK-WRK", 4);
```
Comment (enusbaum) [inline MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs:4865]: I started using the `IntPtr16.Size` const on declarations just to denote what'll be saved there

File: MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs [hunk 3]
```diff
@@ -5056,50 +5027,10 @@ private void fgetc()
...
-                if (string1[i] == string2[i] - 32) continue;
-
-                //1 < 2 == -1 (0xFFFF)
-                //1 > 2 == 1 (0x0001)
-                Registers.AX = (ushort)(string1[i] < string2[i] ? 0xFFFF : 1);
-                return;
-            }
-
-            Registers.AX = 0;
+            Registers.AX = (ushort)(short)String.Compare(string1, string2, /* ignoreCase= */ true);
```
Comment (enusbaum) [inline MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs:5033]: `string` vs `String` alias
`(short)` casting can be removed
You can use parameter naming in C#, so `ignoreCase: true` would work for the 3rd parameter vs. the inline comment

File: MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs [hunk 4]
```diff
@@ -5719,20 +5652,11 @@ private void memcmp()
...
-                if (ptr1Data[i] == ptr2Data[i])
-                    continue;
-
-                if (ptr1Data[i] > ptr2Data[i])
-                    Registers.AX = 0x1;
-                else
-                    Registers.AX = 0xFFFF;
-
-                return;
+                Registers.AX = (ushort)(short)(ptr1Data[i] - ptr2Data[i]);
```
Comment (enusbaum) [inline MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs:5658]: `(short)` casting can be removed

**Review by enusbaum — CHANGES_REQUESTED** [review]

File: MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs (hunk 3 above)
Comment (paladine) [reply MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs:5033]: Cool, using magic numbers like true/false in method calls can be confusing if you're not sure what the parameter is for, but parameter naming takes that ambiguity away

**Pull request merged**